vpc/service/assign_addresses_v3.go

```diff
@@ -98,27 +102,56 @@ type branchENI struct {
 	idx           int
 }
 
+func subnetCacheKey(az, accountID string, subnetIDs []string) string {
+	sort.Strings(subnetIDs)
+	return fmt.Sprintf("az:%s accountID:%s subnetIDs:%s", az, accountID, strings.Join(subnetIDs, ","))
+}
+
 func (vpcService *vpcService) getSubnet(ctx context.Context, az, accountID string, subnetIDs []string) (*subnet, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	ctx, span := trace.StartSpan(ctx, "getSubnet")
-	defer span.End()
+	// NOTE[jigish] caching using the full set of subnetIDs will cause the same subnet to be returned for all hits
+	cacheKey := subnetCacheKey(az, accountID, subnetIDs)
+	r, err, _ := vpcService.getSubnetLock.Do(cacheKey, func() (interface{}, error) {
```

Comment:
I'm thinking that the steps should be:
1. Check if the item is in the cache (and not expired) -- if it is not expired, return immediately
2. Move into a singleflight operation
3. In the single flight operation, check the cache _again_, and see if the subnet is in there (and not expired).
4. Create a (shorter) timeout and try to fetch / renew the subnets
5. If we failed to fetch / renew the subnets, then try to serve the expired version (always, and log a message + mark the span)

Reply:
what do you feel like a better shorter timeout would be? the request timeout is 5 minutes -- so 1 minute?

Reply:
Sure. There's no reason the process of fetching the subnet should be cancelled if the request gets cancelled.

Reply:
timeout added as 1 minute